Orange/widgets/visualize/owmosaic.py

```diff
@@ -438,34 +439,29 @@ def set_data(self, data):
             self.init_combos(None)
             return
 
-        self.color_model.set_domain(self.data.domain)
         self.init_combos(self.data)
 
         self.openContext(self.data)
 
-        # if we first received subset we now call setSubsetData to process it
-        if self.unprocessed_subset_data:
-            self.set_subset_data(self.unprocessed_subset_data)
-            self.unprocessed_subset_data = None
-
-        self.set_color_data()
-
     @Inputs.data_subset
     def set_subset_data(self, data):
-        self.Warning.incompatible_subset.clear()
-        if self.data is None:
-            self.unprocessed_subset_data = data
-            return
-        try:
-            self.subset_data = data.transform(self.data.domain)
-        except:
-            self.subset_data = None
-            self.Warning.incompatible_subset(shown=data is not None)
+        self.subset_data = data
 
     # this is called by widget after setData and setSubsetData are called.
     # this way the graph is updated only once
     def handleNewSignals(self):
-        self.reset_graph()
+        self.Warning.incompatible_subset.clear()
+        self.subset_indices = None
+        if self.data is not None and self.subset_data:
+            try:
+                indices = {e.id for e in self.subset_data.transform(self.data.domain)}
+            except:
+                self.Warning.incompatible_subset(shown=self.subset_indices is not None)
```

Comment:
Maybe `Table.transform` should have an additional argument, a flag that would tell it to raise an exception if no attribute can be transformed. `Table.transform` can probably easily detect this (I haven't checked), it's faster than `np.all(np.isnan(..))` and, most importantly, does not report an incompatible domain when the domain is compatible (maybe even the same), but all data is missing.